### ISSUE TITLE
(AMD-Win/Intel-Lin): clEnqueueReadBuffer (seeds) : CL_INVALID_VALUE and FIXES

### ISSUE THREAD

**Issue by Hyenadae** [minecrafthome/kaktwoos-cl#2]
When attempting to run current Master code, this error occurs in AMD Windows and Intel Linux OpenCL environments.

I've managed to get the code and GPU kernel running on my Vega 56 by faking a seed being found, which sets a **non-zero value** in clEnqueueReadBuffer. 

For some reason, some compilers and environments (Nvidia) do not care the buffer is set to 0, and so the program runs fine. If there is some way to create a buffer while seed_count is equal to 0, while retaining the increase of buffer size for *additional seeds*, then this would be fine. Otherwise, I assume seeds more than X debug seeds set will be lost. 

// Original
       int seed_count = data_out[2];
	// Errors HERE	
        cl_ulong *result = malloc(sizeof(cl_ulong) * seed_count);
        check(clEnqueueReadBuffer(command_queue, seeds, CL_TRUE, 0, **sizeof(cl_ulong) * seed_count**, result, 0, NULL, NULL), "clEnqueueReadBuffer (seeds) ");
		
// Debugged code
    int seed_count = data_out[2];
        seed_count = 1;
        seedbuffer_size = sizeof(cl_ulong) * seed_count;
        cl_ulong *result = malloc(sizeof(cl_ulong) * seed_count);

	check(clEnqueueReadBuffer(command_queue, seeds, CL_TRUE, 0, seedbuffer_size, result, 0, NULL, NULL), "clEnqueueReadBuffer2 (seeds) ");
 
// End of erroring section

Strangely enough, data_out appears to work fine when the pre-set buffer is working. I got a seed during a test run, but I have to verify it:

Found seed: 2080652165, 8936333004391317531, height: 27

**Comment by Hyenadae** [comment]
The cleaner, stupidest possible "fix" code is this: (Keeping everything else the same)

 seedbuffer_size = sizeof(cl_ulong) + sizeof(cl_ulong) * seed_count;
 cl_ulong *result = malloc(sizeof(cl_ulong) + sizeof(cl_ulong) * seed_count);

check(clEnqueueReadBuffer(command_queue, seeds, CL_TRUE, 0, seedbuffer_size, result, 0, NULL, NULL), "clEnqueueReadBuffer (seeds) ");


Basically, set the default buffer size to *1* seed's worth, then have it increase by the proper multiple after real seeds are found. For some reason this **isn't** required on OCL-Nvidia to work without a segfault or this issue's error.

**Comment by Hyenadae** [comment]
I can confirm Intel Beginet (Ubuntu 2.0CL - v1.3.2) calculates two seeds properly using the debug input for Height 20 cactus. 

kaktwoos-cl.bin -d 0 -s 0 -e 10000000000 -cs 9567961692053 -n1 856 -n2 344 -n3 840 -di 0 -ch 12

Output: 
    Found seed: 5861434992872476197, 77849775653, height: 20
5861434992872476197
Speed: (n/a)m/s, 63.49%, (n/a).0s
    Found seed: 5861435033067469701, 118044769157, height: 20
5861435033067469701

Found seeds: 
    5861434992872476197
    5861435033067469701

AMD's outputs and calculated seeds are still very wrong: 
  Found seed: 540364325, 2320847105164836882, height: 18
540364325

Found seed: 2080652165, 8936333004391317531, height: 27
2080652165

Processed **1410065408** seeds in **-338401648210390791602766548924559975396556347317060244479916587760885010895121993943417742520550821222578684805443157300312508286372565495573425368294545205631238063796405795762761474691668818880891928453131949261271768816895236984623071232.000000 seconds

Found seeds:
    540364325
    2080652165**

**Comment by Hyenadae** [comment]
I think I did it. Kaktwoos should now run and print valid seeds on both Intel and AMD:OpenCL now. 

Benchmarks with fixes:
1645MHz : 930Mhz Vega 56 OC= 25.95mseed/s 
Core i3-7100U (Intel U620) = 2.90mseed/s  (Estimated)

On way slower systems, there will essentially be no output or accurate statistics. I did a run of a smaller amount of seeds instead and compared.  (40.5 seconds vs 4.5 on AMD with known rate)

Code Fixes Part 2:
(Change output text's %lu to SCNd64 as below.)

     for (int i = 0; i < seed_count; i++) {
            printf("    Found seed: %"SCNd64 ", %llu, height: %d\n",
                    result[i],
                    result[i] & ((1ULL << 48ULL) - 1ULL),
                    (int)(result[i] >> 58ULL));
            fprintf(stderr, "%"SCNd64 "\n", (cl_ulong)result[i]);
            found_seeds[total_seed_count++] = result[i];
        }

```
printf("Done\n");
    printf("Processed %" SCNd64 " seeds in %f seconds\n",
            end - start,
            (double)(end_time - start_time) / CLOCKS_PER_SEC);
    printf("Found seeds: \n");
    for (int i = 0; i < total_seed_count; i++) {
        printf("    %"SCNd64 "\n", found_seeds[i]);
    }

```

--------------------------------------------------------------------------------------------------------------

Windows AMD-OpenCL Requirements:
**64bit GCC (9.2.0)** (I used TDM-GCC-64, but the main thing is doing the linking properly)

https://github.com/GPUOpen-LibrariesAndSDKs/OCL-SDK/releases installed. 

gcc .\main.c -o main.exe -LC:\Path\to\AMD-OCL\lib\x86_64\ -lOpenCL

Linux Intel OpenCL Requirements:
Beginet packages and beginet-dev installed
GCC (64bit, 8.2 or newer)
compiled with gcc main.c -o main.bin -lOpenCL

Both tested with:
./main.exe -d 0 -s 0 -e 10000000000 -cs 9567961692053 -n1 856 -n2 344 -n3 840 -di 0 -ch 12
and verified with kaktwoos.jar validator

Probably a good idea to manually copy the AMD Windows OpenCL libraries to mingw lib and include folder, and hoping they're detected. I did a bunch of mangling on my end to the code to before, but this method should work as well.
